### ISSUE TITLE
[IDEA] Enable listing of all controlpanel settings

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#7117]
**Problem**: The Controlpanel design is not good. It is difficult to locate settings. It is not searchable. It is difficult to modify the controlpanel because of multiple tab levels, at arbitrary places, *and* in arbitrary combination with hard coded content. For generated content, the filters are embedded in the tiddler text so the filter cannot be extracted for reuse. The whole controlpanel also has an archaic appearance.

**Describe the solution you'd like**
I want a list of *all* settings seen in the Controlpanel. I've long pondered over the problems mentioned above and concluded that some way to *list all settings* is the **minimum requirement** for doing further work with the items, such as creating custom controlpanels, searchability, etc.

I am not sure *how* this should be implemented, but it seems the most direct way would be a common tag on all relevant tiddlers. If this is considered cluttery, then instead a common *field*.

**Describe alternatives you've considered**
An alternative to a single *common* tag or field, would be a tag or field build up with name spaces but this would (1) require more work and (2) still be arbitrary. One neat aspect then could be that shared name spaces could be used to auto-select e.g template for how to present the setting.

**Other notes**
There are many other systems of greater complexity that feature better controlpanels, such as various OS and browsers - and they don't use a static Controlpanel like TW.

**Comment by twMat** [comment]
@pmario - you bring up the problem with findability for translations. My OP would definitely simplify to solve this problem, but your problem is very specific where my request is much more general. I want to give the end user access to manipulate and hack the controlpanel in a *general* way. As noted, I've concluded that the first thing needed is a list of the titles of the tiddlers that are displayed as settings in the controlpanel. (I've updated the OP to clarify this.)

Let me clarify by using your quote:

>I personally would start with: https://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FControlPanel%2FSettings .. If someone is able to make that list searchable with a different language than English, I think there is a good chance to get the rest going too.

Those items *already* fulfil most of what I'm requesting: They *are* already listable via `[all[shadows+tiddlers]tag[$:/tags/ControlPanel/Settings]]`, so a user can easily access them for further manipulation. But for a user to control the full controlpanel, we'd need *all* equivalent filters for the various controlpanel tabs. But there are several problems because the various settings in the tabs are inconsistently presented: Many controlpanel tabs don't at all display such filter generated settings, but instead hard coded content. And when they *do* display filter generated settings, like in the Settings tab you refer to, then the filter is *embedded* in the code so it can't be programmatically used. Yes, it is possible to *construct* the filter out of the tiddler title (i.e `$:/core/ui/ControlPanel/Settings`) but only for some of the tabs so it is not consistent.

>The filter "[all[shadows+tiddlers]prefix[$:/config]]" returns 248 matches. Most of them can probably be found in the control panel. I don't know. ... IMO every single one of them would need some meta-data, that connects the value to a section in the control-panel. ...

This is starting in the wrong end. Those are the setting *values*. But I want the end user to be able to customize the settings UI. 

You bring up the specific example of findability in *translations*; For end users, I think the *only* solution is to enable adding of custom meta-data. Maybe I never remember the term "camel case" but term "autolink" always pops up in my head when I try to find the relevant setting. So a good Controlpanel UI should simplify for the user to add such metadata, and without overwriting core tids. This is also a case in point for my OP and why we need a *list* of all controlpanel settings. Adding custom metadata would be simple (...well) then.

**Comment by kookma** [comment]
@twMat 
There is some old discussion on making settings searchable, among them see and the suggestion by @saqimtiaz 

- https://github.com/Jermolene/TiddlyWiki5/issues/4957
- https://github.com/Jermolene/TiddlyWiki5/issues/4507

Other threads
- https://github.com/Jermolene/TiddlyWiki5/issues/1690
- https://github.com/Jermolene/TiddlyWiki5/issues/1689

**Comment by Jermolene** [comment]
It's worth breaking down the precise level of backwards compatibility that we think is appropriate. For example, these constraints get progressively looser:

* Do we remain compatible with existing wikitext that navigates to a specific tab of the control panel by manipulating the state tiddlers?
* Do we remain compatible with existing wikitext that transcludes or otherwises reuses core control panel tabs by tag? Or do we go further and preserve the titles so that direct transclusions continue to work?
* Do we remain compatible with plugins that add new control panel tabs with one of the existing tags? 

When you think of it like that it is clear that we wouldn't be able to make any worthwhile improvements while retaining total backwards compatibility.

Therefore, changes like this are going to break backwards compatibility, and the question for us is one of degree.

My view is that the control panel is in sufficiently desperate need of an overhaul that we should consider going for minimal backwards compatibility, which would mean just observing the last constraint in the list above: just to make sure that if a plugin provides old school control panel tabs that they are available somewhere within the control panel. The presentation of those tabs doesn't need to be compatible with the new fangled control panel entries. They can just be pushed together in a legacy panel.

**Comment by pmario** [comment]
I personally do like the tab structure. I don't care how it is created, but I think there needs to be a tab-like structure, because otherwise it will be impossible to present ALL settings with a clean UI. 

As a user I want to be able to see a clear structure without limiting the view with a search term. 

I think it's possible to create a "control" namespace and dynamically create tabs out of that namespace. 

But if every setting gets a new name, we will need to create a completely new UI, since all the tiddler titles which are used in the control panel are usually hardcoded in a similar way in the UI.

**Comment by pmario** [comment]
>The obvious alternative to tabs would be to use an expandable/collapsible tree/TOC to the left of the control pane content.

Yes .. I did think about that too. I was thinking about an [TW Explorer type of view](https://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FMoreSideBar%2FExplorer) 

As I wrote a TOC needs tags. Not every setting has a tiddler until the user changes the setting. So by default there is nothing to tag. .. OR we will need to make every setting in a similar way as button-visibility settings. Which I personally would consider overhead aka bloat.

**Comment by AnthonyMuscio** [comment]
I think we could benefit from a search tool that can simply begiven a filter and it searches those tiddlers and their fields only, a quick check of some of the control panel tabs shows they are consistently named with the prefix $:/core/ui/ControlPanel/ except for non core items made visible using a set of tags by third party solutions/plugins, these tags can be part of the filter.
* Further an option could be built to first render each tiddler before searching within it and returning the list of tiddlers where the search string exists after render. We could call this "full search"
* I would then include this on the control panel page for a comprehensive search of all control panel tiddlers.
* If we make this a core macro user can use it elsewhere to the same effect, including the full search option.

I have long thought a set of menus - looking like breadcrumbs and somewhat like the tree macro would help with navigation.

None the less I recognise the issue raised by @twMat but would add the issue goes further the tiddlers even just because of their title such as control panel, tiddler manager and tag manager do not come up in standard search results as they are system tiddlers.
* What about a plugin of regular tiddlers designed to make the control panel and system tiddlers with readable content more searchable/accessible? Ideally with translations.
* A Key feature would be exactly what Mat asks for.

**Comment by pmario** [comment]
>...tiddler manager and tag manager do not come up in standard search results as they are system tiddlers.

I think that's an easy fix. Just let us create 2 core tiddlers 1st "Tiddler Manager" 2nd "Tag Manager" and transclude the shadow tiddlers into them. 

**Suggestion 1**
So languages that have their own edition will be able to translate the titles and we are done. 

**Alternative Suggestion**
We can add a link to those 2 ore even more important tiddlers like the ControlPanel to the GettingStarted tiddler, if we don't want to create new tiddlers that would be part of empty.html

@Jermolene What do you think about that?

**Comment by Jermolene** [comment]
> > ...tiddler manager and tag manager do not come up in standard search results as they are system tiddlers.
> 
> I think that's an easy fix. Just let us create 2 core tiddlers 1st "Tiddler Manager" 2nd "Tag Manager" and transclude the shadow tiddlers into them.

I do not think it is desirable for control panel entries to show up in the standard sidebar search. TiddlyWiki goes to quite a lot of trouble to only present tiddlers that the user has created themselves; it is intentional that one has to explicitly look for shadows or system tiddlers.

One could imagine showing the control panel search results in an additional tab but I would resist putting that into the core because it makes the search results significantly more complex to navigate when the tabs are enabled.

In any event, from the users perspective it seems logical that they should do these searches within the context of the control panel.

**Comment by pmario** [comment]
We are a bit OT here. Sorry Mat. 

>I do not think it is desirable for control panel entries to show up in the standard sidebar search.

Sorry, I haven't been very clear. 

I think the problem is that [$:Manger](https://tiddlywiki.com/#%24%3A%2FManager) and [$:/TagManager](https://tiddlywiki.com/#%24%3A%2FTagManager) do not show up anywhere. 

They are not part of the ControlPanel and they cannot be found with the search. So without specifically searching for them, you don't find them. .. So it makes it very hard for new users to learn about them at all. 

That's why I did suggest that we either mention them somewhere or we create 2 new tiddlers without the system prefix.

**Comment by Jermolene** [comment]
> I think the problem is that [$:Manger](https://tiddlywiki.com/#%24%3A%2FManager) and [$:/TagManager](https://tiddlywiki.com/#%24%3A%2FTagManager) do not show up anywhere in tiddler content
> 
> They are not part of the ControlPanel and they cannot be found with the search. So without specifically searching for them, you don't find them. .. So it makes it very hard for new users to learn about them at all.

That makes sense, and I agree that things could be improved. Perhaps it's worth a separate ticket because it's not really related to the control panel. My thought was to gather all the tools that we provide for the user to work with their own data into a revised, tabbed `$:/Manager` tiddler. It would be analogous to the existing control panel, but concerned with the users data rather than with system configuration. But that still doesn't solve the problem of giving the user clear and obvious signposts to find it. An obvious solution would be to display the `$:/Manager` toolbar button by default. But that would mean breaking the rule of only having three icons shown by default in any toolbar.

**Comment by pmario** [comment]
I found this [ControlPanel overview](https://scrollback.tiddlyhost.com/#%24%3A%2Fplugins%2FHackSearch%2Ftab%2FControlPanel:%24%3A%2FAdvancedSearch%20%24%3A%2Fplugins%2FHackSearch%2Ftab%2FControlPanel) from @twMat, which seems to be dynamically created. 

IMO it could be a basis to list all options from all sub-panels

![image](https://user-images.githubusercontent.com/374655/218267761-8aed7cfc-d18a-420a-ba60-ef5cda2ff62d.png)

**Comment by AnthonyMuscio** [comment]
Yes twMat solution to this is really good, especially being dynamically created.

My only desire is that a more general solution is made available, so we can use it for the sidebar and users can do so with their own content. This includes being able to search rendered content in a tree of tabs/tiddlers.

**Comment by pmario** [comment]
@twMat -- Would this be nice?

![image](https://github.com/user-attachments/assets/67a4715d-e605-4f42-9f45-9e05b464dceb)

**Comment by yaisog** [comment]
> Therefore, changes like this are going to break backwards compatibility, and the question for us is one of degree.

Hi @Jermolene, might it make sense to retain _$:/ControlPanel_ for backwards compatibility and create a new _$:/Settings_ tiddler that implements the UI changes / TOC / search? Both would act on the same set of config tiddlers and could be used interchangingly. The meta information that needs to be added to connect lingo tiddlers to config settings that would be needed for the latter doesn't need to interfere with the former.

At some code cleanup point in the future, _$:/ControlPanel_ might then be moved to a plugin.

**Comment by pmario** [comment]
The problem with the control panel is the search function not the display

**Comment by yaisog** [comment]
Now I am confused. The post that I quoted and replied to was talking about an overhaul of the settings presentation and how third-party content that relied on the $:/ControlPanel structure would be affected, was it not?

Concerning searchability, if we were to add a `caption` field to each relevant config tiddler that contained the description, either as a string or transclusion or `<<lingo>>` macro, it would be simple to implement a search of all settings. It would also be backwards compatible, as these fields are otherwise ignored. It is a lot of tedious work, though, and will probably get stuck by our usual inability to find an agreement on minor details as did e.g. #8997.

**Comment by kookma** [comment]
I like the @yaisog idea! The $:/Setting with searchability in caption ad title, makes sense. Great idea!
